src/Features/Core/Portable/RemoveAsyncModifier/AbstractRemoveAsyncModifierCodeFixProvider.cs

```diff
@@ -0,0 +1,236 @@
+        }
+
+        private void RemoveAsyncModifier(SyntaxEditor editor, SemanticModel semanticModel, SyntaxNode originalNode, IMethodSymbol methodSymbol, KnownTypes knownTypes)
+        {
+            // SyntaxEditor can't change or remove tokens so we have to replace the whole node for the method
+            // which means this edit must come first
+            var replacementNode = RemoveAsyncModifier(methodSymbol, originalNode, knownTypes);
+            editor.ReplaceNode(originalNode, replacementNode);
+
+            // Now that we've replaced the original node we have to work on the replacement for future work
+            if (TryGetExpressionBody(replacementNode, out var expressionBody))
+            {
+                if (methodSymbol.ReturnType == knownTypes._taskType)
+                {
+                    // We need to add a return so we have to convert to a block body
+                    var blockBodiedNode = ConvertToBlockBody(replacementNode, expressionBody, editor);
+
+                    // Expression bodied members can't have return statements so if we can't convert to a block
+                    // body then we've done all we can
+                    if (blockBodiedNode != null)
+                    {
+                        editor.ReplaceNode(replacementNode, blockBodiedNode);
+                        // We need to get the block inside the block bodied method to know where to add the return
+                        var node = GetLastChildOfBlock(blockBodiedNode);
+                        AppendTaskCompletedTaskReturn(node, editor, knownTypes);
+                    }
+                }
+                else
+                {
+                    // For Task<T> returning expression bodied methods we can just wrap the whole expression
+                    WrapExpressionWithTaskFromResult(expressionBody, editor, knownTypes);
+                }
+            }
+            else
+            {
+                // Block bodied methods might have return statements so update them to task returning
+                ChangeReturnStatements(replacementNode, editor, knownTypes);
+
+                // An "async Task" method has an implicit return at the end, so removing async means
+                // we need to insert it explicitly if the end of the method is reachable.
+                if (methodSymbol.ReturnType == knownTypes._taskType)
+                {
+                    // We have to use the original node to do control flow analysis, but the reachability of it is the same
+                    var controlFlow = AnalyzeControlFlow(semanticModel, originalNode);
+
+                    if (controlFlow != null)
+                    {
+                        // For local functions and block bodied lambdas the EndPointIsReachable is false but we still might need to
+                        // insert a return. We can tell by checking for the presence of exit points that aren't return statements
```

Comment:
hrmm... waht are the exit points in those cases?

Reply:
This was probably my fault. I reworked the code to be language agnostic and am analyzing control flow between the start and end statements within the block, and the extra condition is unncessary.